torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -70,6 +71,29 @@ std::vector<c10::DeviceIndex> getDevicesForTensors(
   }
 }
 
+std::vector<at::cuda::CUDAStream> getStreamsForTpMessage(
+    const tensorpipe::Message& tpMessage) {
+  std::vector<at::cuda::CUDAStream> streams;
+  for (auto& tensor : tpMessage.tensors) {
+    if (!tensor.metadata.empty()) {
+      streams.emplace_back(at::cuda::getStreamFromPool(
+          c10::DeviceIndex(std::stoi(tensor.metadata))));
+    }
+  }
+  return streams;
+}
+
+// use unique_ptrs as CUDAStreamGuard is not movable or copyable
```

Comment:
Using guards seems a bit contrived, and probably not the way they were meant to be used (it seems they delete their copy/move constructors on purpose). Would it make sense to instead just call `at::cuda::setCurrentCUDAStream(stream);` for each stream? This is what the Python `torch.cuda.stream(...)` context manager does.

Reply:
That will work. Or maybe I can add a batch stream guard for this purpose, it might be safer to use a guard.

Reply:
The "current" stream is a thread local variable for the thread in which it is set, right? We only set CUDA streams ad current from the threads in our own thread pool, and we do so anew for each new remote function, thus overwriting the previous current streams. So I think it should be fine to not use a guard, because there is never a need to "unset" a stream and return to the one that was set before. Right?

Reply:
streams are also set in pipe->read/write callback functions. If we have the guarantee that all cuda ops in TensorPipe's event loop are explicitly scoped to the correct stream, then we can save the unset on the agent side. But I am not sure if that will always be the case. If this assumption is no longer true in the future, it will be really hard to debug.

Reply:
> If we have the guarantee that all cuda ops in TensorPipe's event loop are explicitly scoped to the correct stream

Yes, that's something that TensorPipe will guarantee, it will only use the CUDA stream that is implicitly passed to it, it will not use any current/default CUDA stream. (Also, the concept of "current" stream is a thread_local variable in PyTorch, so TensorPipe doesn't even have access to it).

Thus I think we shouldn't set streams as current in read/write callbacks, but only just before calling into user code, and that always happen inside a worker thread. Therefore there is no need to re-set the previous stream, because the next job run on those threads will set the current streams anew.